File: nodejs/service/like.service.js
const likeModel = require("../model/like.model");

class LikeService {
    reactPost = async (req) => {
        const postId = req.params.postId;
        const userId = req.user.id;
        const { reaction } = req.body;
        if (!["like", "dislike"].includes(reaction)) {
            throw new Error("Invalid Reaction");
        }
        let existing = await likeModel.findOne({
            userId,
            postId,
        });
        if (!existing) {
            existing = await likeModel.create({
                userId,
                postId,
                reaction,
            });
        } else {
            if (existing.reaction === reaction) {
                await likeModel.deleteOne({
                    _id: existing._id,
                });
            } else {
                existing.reaction = reaction;
                await existing.save();
            }
        }
        return await this.getReactions(req);
    };
    getReactions = async (req) => {
        const postId = req.params.postId;
        const userId = req.user.id;
        const likes = await likeModel.countDocuments({
            postId,
            reaction: "like",
        });
        const dislikes = await likeModel.countDocuments({
            postId,
            reaction: "dislike",
        });
        const mine = await likeModel.findOne({
            postId,
            userId,
        });
        return {
            likes,
            dislikes,
            myReaction: mine ? mine.reaction : null,
        };
    };
}
module.exports = new LikeService();

File: DJANGO/api/services/like_service.py
from api.models import Like, Post


class LikeService:

    @staticmethod
    def react_post(post_id, reaction, user):

        if reaction not in ["like", "dislike"]:
            raise Exception("Invalid Reaction")

        post = Post.objects.get(id=post_id)

        existing = Like.objects.filter(
            user=user,
            post=post
        ).first()

        if not existing:

            Like.objects.create(
                user=user,
                post=post,
                reaction=reaction
            )

        else:

            if existing.reaction == reaction:
                existing.delete()

            else:
                existing.reaction = reaction
                existing.save()

        return LikeService.get_reactions(post_id, user)

    @staticmethod
    def get_reactions(post_id, user):

        post = Post.objects.get(id=post_id)

        likes = Like.objects.filter(
            post=post,
            reaction="like"
        ).count()

        dislikes = Like.objects.filter(
            post=post,
            reaction="dislike"
        ).count()

        mine = Like.objects.filter(
            post=post,
            user=user
        ).first()

        return {
            "likes": likes,
            "dislikes": dislikes,
            "myReaction": mine.reaction if mine else None
        }